arangod/Agency/Agent.h

```diff
@@ -289,12 +293,6 @@ class Agent : public arangodb::Thread,
     _appendCV.broadcast();
```

Comment:
I think the code is correct as it stands. Here is my reasoning: 

Under Unix we use pthread condition variable, under Windows we use CONDITION_VARIABLE.

Both say explicitly in the docs that the broadcast may be called under the lock or outside (references: http://pubs.opengroup.org/onlinepubs/009695399/functions/pthread_cond_broadcast.html for pthreads and https://msdn.microsoft.com/de-de/library/windows/desktop/ms682052(v=vs.85).aspx for Windows).

So let's argue for "not under the lock":

In this particular case there is an additional boolean variable `_agentNeedsWakeup` which indicates whether or not the agent thread needs to wake up. It is crucial that this is only modified and read under the mutex (which `Agent::wakeupMainLoop` does). We need to show that it can never happen that the broadcast is done whilst the agent thread is still awake, such that the broadcast evaporates unheard and then the agent thread goes to sleep with `wait`. Well, this is what `_agentNeedsWakeup` and the Mutex ensure: `Agent::run` first acquires the mutex (`Agent.cpp:1229`), then checks the bool and only if that is `false` it calls - under the mutex, `wait`. `Agent::wakeupMainLoop` acquires the Mutex, sets the bool to `true` under the mutex, releases the mutex and then calls broadcast.

The argument is that either `wakeupMainLoop` gets the mutex first but then `run` sees the bool as `true` and does not go to sleep, or `run` gets the mutex first but then it goes to sleep and releases the mutex in one atomic operation (guarantees of `wait`), that is, `wakeupMainLoop` only sets the flag to `true` after that and thus also calls broadcast after the main thread has gone to sleep (since broadcast "happens after" the block with the mutex.

Finally, no memory barriers are needed for the access to `_agentNeedsWakeup` because of the mutex and Posix (and almost certainly also Windows) guarantees full memory barriers for any kind of explicit thread synchronization.

I studied this type of situation some time ago and I found sources where it was not only suggested that the code is correct but even better. The reason for that is that if one calls the broadcast (or indeed signal) under the mutex it can happen that it wakes up a sleeping thread, only to go back to sleep to wait to acquire the mutex! Recall that `wait` guarantees that waking up and reacquiring the mutex is an atomic operation. I have no evidence that this is true, but I find it plausible.